clang/lib/CIR/CodeGen/CIRGenBuilder.h

```diff
@@ -854,24 +855,36 @@ class CIRGenBuilderTy : public CIRBaseBuilderTy {
 
     mlir::Type SubType;
 
+    auto getOffsetAndIndex =
+        [](int64_t Offset, int64_t EltSize) -> std::pair<int64_t, int64_t> {
+      int64_t DivRet = Offset / EltSize;
+      if (DivRet < 0)
+        DivRet -= 1; // make sure offset is positive
+      int64_t ModRet = Offset - (DivRet * EltSize);
+      return {DivRet, ModRet};
+    };
+
     if (auto ArrayTy = mlir::dyn_cast<mlir::cir::ArrayType>(Ty)) {
-      auto EltSize = Layout.getTypeAllocSize(ArrayTy.getEltType());
-      Indices.push_back(Offset / EltSize);
+      int64_t EltSize = Layout.getTypeAllocSize(ArrayTy.getEltType());
       SubType = ArrayTy.getEltType();
-      Offset %= EltSize;
+      auto const [Index, NewOffset] = getOffsetAndIndex(Offset, EltSize);
+      Indices.push_back(Index);
+      Offset = NewOffset;
     } else if (auto PtrTy = mlir::dyn_cast<mlir::cir::PointerType>(Ty)) {
-      auto EltSize = Layout.getTypeAllocSize(PtrTy.getPointee());
-      Indices.push_back(Offset / EltSize);
+      int64_t EltSize = Layout.getTypeAllocSize(PtrTy.getPointee());
       SubType = PtrTy.getPointee();
-      Offset %= EltSize;
+      auto const [Index, NewOffset] = getOffsetAndIndex(Offset, EltSize);
+      Indices.push_back(Index);
+      Offset = NewOffset;
```

Comment:
Could you please add tests that cover these two branches as well?

Reply:
How to add test for PointerType in global. I think for quite long time but no idea?

Reply:
The pointer branch looks indeed impossible to reach. Maybe we could delete it and let it asserts.